File: track/core/time_entry_list.py
from dataclasses import dataclass
from typing import List

from track.core import TimeEntry


@dataclass
class TimeEntryList:
    values: List[TimeEntry]

    def get_total_duration(self):
        total_duration = 0
        for entry in self.values:
            total_duration += entry.duration
        return total_duration

    def group_entries_by_description(self):
        grouped_entries = {}
        for entry in self.values:
            if entry.description in grouped_entries:
                grouped_entries[entry.description].duration += entry.duration
            else:
                grouped_entries[entry.description] = entry
        return grouped_entries.values()

File: src/core/entities/time-entry-list.ts
import { Duration } from "./duration";
import {TimeEntry} from "./time-entry";

export class TimeEntryList {
  values: Array<TimeEntry>;

  constructor(values: Array<TimeEntry>) {
    this.values = values;
  }

  getTotalDuration(): Duration {
    const totalDuration = new Duration(0)

    for (const {duration} of this.values) {
      totalDuration.add(duration)
    }

    return totalDuration
  }

  groupEntriesByDescription(): TimeEntryList {
    const groupedEntries: Record<string, TimeEntry> = {};

    for (const entry of this.values) {
      const dictId = entry.project.id + entry.description
      if (groupedEntries[dictId]) {
        groupedEntries[dictId].duration.add(entry.duration);
      } else {
        groupedEntries[dictId] = new TimeEntry(entry)
      }
    }

    return new TimeEntryList(Object.values(groupedEntries));
  }
}
